### PR TITLE
cache http probe request object

### LINKED ISSUE #115939: Reuse the http request object for http probes
Today we construct the request object for every http request for the http probes. This involves concatenating some strings, setting headers, etc.

Probes are simple http get requests that are executed synchronously (for a single probe) and has no request body.

The optimization may be to reuse the request object for all probe executions. This likely save some cpu and memory for the kubelet process.

Some code references:
- [DoProbe](https://github.com/kubernetes/kubernetes/blob/4dfd2fe74d96f67c9752bbdc9ae9983a6e5f0492/pkg/probe/http/http.go#L93) is executed today on each probe execution.
- New request are [being created](https://github.com/kubernetes/kubernetes/blob/4dfd2fe74d96f67c9752bbdc9ae9983a6e5f0492/pkg/kubelet/prober/prober.go#L146) for each execution.
- New object will need to be stored in [worker](https://github.com/kubernetes/kubernetes/blob/4dfd2fe74d96f67c9752bbdc9ae9983a6e5f0492/pkg/kubelet/prober/worker.go#L40) that for http probes will hold the http request object. This object will need to encapsulate the logic of [runProbe](https://github.com/kubernetes/kubernetes/blob/4dfd2fe74d96f67c9752bbdc9ae9983a6e5f0492/pkg/kubelet/prober/prober.go#L138) method of the prober.

I'm marking this issue as a good first issue, as the issue doesn't require deep knowledge of Kubernetes. However it is not an easy issue, a few parts of code will be affected and change needs to be done very carefully.

***Please review code and make sure you understand the task before self-assigning the issue.***

Implementation notes

1. If you can send the small benchmark results on probes with and without this optimization, it will be a good starting point. Note, for the benchmark, each probe can run once a second, not more often. But the number of probes is virtually unlimited since we are not defining the limit on number of containers per Pod. Realistically we are talking about 110 pods (max) with maybe 2 containers each, with 3 probes each (660 probes).
2. Code change may be big. So make sure to help reviewers to review the code. One idea may be to split implementation into multiple commits to simplify the review. First, creating the object encapsulating the logic of `runProbe`, second storing and initializing this object in `worker`, and third, actual reusing of the request object.

/sig node
/good-first-issue
/help-wanted
/area kubelet

### PR DESCRIPTION
**Pull request by Taoup**
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
Http probe requests are constructed for every http probe that's being executed, which is probably a waste of cpu cycles and memory. This PR uses sync.Map as a cache to reuse the constructed http requests.

#### Which issue(s) this PR fixes:

Fixes #115939 

#### Special notes for your reviewer:
The original PR #116058 has been closed and not active for quite some time, so I'd like to present my implementation for discussion. 

Besides, I'm new to Kubernetes and the golang overall.  I'd like to submit is PR asking for help  to check the sanity of my implementation before moving on to next step, i.e. benchmarking and so on. 

#### Does this PR introduce a user-facing change?

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by Taoup** [comment]
update benchmark result:

```
benchstat old.txt new.txt 
goos: linux
goarch: amd64
pkg: k8s.io/kubernetes/pkg/kubelet/prober
cpu: Intel(R) Xeon(R) Gold 5220R CPU @ 2.20GHz
        │   old.txt   │               new.txt               │
        │   sec/op    │   sec/op     vs base                │
Probe-8   557.6m ± 5%   164.7m ± 6%  -70.47% (p=0.000 n=10)

        │   old.txt    │               new.txt                │
        │     B/op     │     B/op      vs base                │
Probe-8   359.9Mi ± 0%   142.2Mi ± 0%  -60.49% (p=0.000 n=10)

        │   old.txt    │               new.txt               │
        │  allocs/op   │  allocs/op   vs base                │
Probe-8   3466.0k ± 0%   693.2k ± 0%  -80.00% (p=0.000 n=10)
```

**Comment by Taoup** [comment]
> update benchmark result:
> 
> ```
> benchstat old.txt new.txt 
> goos: linux
> goarch: amd64
> pkg: k8s.io/kubernetes/pkg/kubelet/prober
> cpu: Intel(R) Xeon(R) Gold 5220R CPU @ 2.20GHz
>         │   old.txt   │               new.txt               │
>         │   sec/op    │   sec/op     vs base                │
> Probe-8   557.6m ± 5%   164.7m ± 6%  -70.47% (p=0.000 n=10)
> 
>         │   old.txt    │               new.txt                │
>         │     B/op     │     B/op      vs base                │
> Probe-8   359.9Mi ± 0%   142.2Mi ± 0%  -60.49% (p=0.000 n=10)
> 
>         │   old.txt    │               new.txt               │
>         │  allocs/op   │  allocs/op   vs base                │
> Probe-8   3466.0k ± 0%   693.2k ± 0%  -80.00% (p=0.000 n=10)
> ```

This benchmark does probes in serial mode, which I thought is way more easier to implement than a parallel one. I'm not sure if this result is representative enough. Require for more comments, :)

**Comment by Taoup** [comment]
Update: parallel benchmark result
```
root ➜ /workspaces/kubernetes (master) $ benchstat old.txt new.txt 
goos: linux
goarch: amd64
pkg: k8s.io/kubernetes/pkg/kubelet/prober
cpu: Intel(R) Xeon(R) Gold 5220R CPU @ 2.20GHz
        │    old.txt    │               new.txt                │
        │    sec/op     │    sec/op     vs base                │
Probe-8   173.10m ± 19%   21.70m ± 19%  -87.46% (p=0.000 n=10)

        │    old.txt    │               new.txt                │
        │     B/op      │     B/op      vs base                │
Probe-8   251.30Mi ± 0%   33.66Mi ± 0%  -86.61% (p=0.000 n=10)

        │   old.txt    │               new.txt               │
        │  allocs/op   │  allocs/op   vs base                │
Probe-8   3367.5k ± 0%   594.7k ± 0%  -82.34% (p=0.000 n=10)
```

the code doing this benchmark:
https://github.com/Taoup/kubernetes/commit/853845791b5f8ab1ce36ff113f9e433f2ec4c343

File: pkg/kubelet/prober/prober.go [hunk 1]
```diff
@@ -142,9 +147,27 @@ func (pb *prober) runProbe(ctx context.Context, probeType probeType, p *v1.Probe
...
+		} else {
+			req = reqAny.(*http.Request)
+		}
+		// cache hit, but pod IP changed, recreate the request object
+		if p.HTTPGet.Host == "" && req.URL.Hostname() != status.PodIP {
+			req, err = httpprobe.NewRequestForHTTPGetAction(p.HTTPGet, &container, status.PodIP, "probe")
+			if err != nil {
+				return probe.Unknown, "", err
+			}
+			httpProbeReqCache.Store(key, req)
```
Comment (bart0sh) [inline pkg/kubelet/prober/prober.go:170]: Do I understand correctly that in case of cache miss this is repeated twice?
```
req, err = httpprobe.NewRequestForHTTPGetAction(p.HTTPGet, &container, status.PodIP, "probe")
if err != nil {
	return probe.Unknown, "", err
}
httpProbeReqCache.Store(key, req)
```

Would it make sense to combine `!found` and this scope?

Comment (Taoup) [reply pkg/kubelet/prober/prober.go]:
```
Would it make sense to combine !found and this scope?
```
Done

File: pkg/kubelet/prober/prober.go [hunk 2]
```diff
@@ -40,6 +42,9 @@ import (
 
 const maxProbeRetries = 3
 
+// Requests cache for http probe
+var httpProbeReqCache sync.Map
```
Comment (aojea) [inline pkg/kubelet/prober/prober.go:46]: why a global and not part of the prober struct?

File: pkg/kubelet/prober/prober.go [hunk 3]
```diff
@@ -183,6 +197,13 @@ func (pb *prober) runProbe(ctx context.Context, probeType probeType, p *v1.Probe
 	}
 }
 
+func cachedRequestStaled(p *v1.Probe, req *http.Request, status v1.PodStatus) bool {
+	if p.HTTPGet.Host == "" && req.URL.Hostname() != status.PodIP {
+		return true
+	}
+	return false
```
Comment (bart0sh) [inline pkg/kubelet/prober/prober.go:204]:
```suggestion
	return p.HTTPGet.Host == "" && req.URL.Hostname() != status.PodIP {
```

Considering the suggestion, do we need a separate function for this?

Comment (Taoup) [reply pkg/kubelet/prober/prober.go:204]: A separate function would help with readability and make the original if statement not too verbose. without it, it will look like this.
```go
		reqAny, found := pb.httpProbeReqCache.Load(key)
		if !found || (p.HTTPGet.Host == "" && reqAny.(*http.Request).URL.Hostname() != status.PodIP) {
			req, err = httpprobe.NewRequestForHTTPGetAction(p.HTTPGet, &container, status.PodIP, "probe")
			if err != nil {
```

File: pkg/kubelet/prober/prober.go [hunk 4]
```diff
@@ -142,9 +147,18 @@ func (pb *prober) runProbe(ctx context.Context, probeType probeType, p *v1.Probe
...
+		reqAny, found := httpProbeReqCache.Load(key)
+		if !found || cachedRequestStaled(p, reqAny.(*http.Request), status) {
+			req, err = httpprobe.NewRequestForHTTPGetAction(p.HTTPGet, &container, status.PodIP, "probe")
+			if err != nil {
+				return probe.Unknown, "", err
+			}
+			httpProbeReqCache.Store(key, req)
+		} else {
+			req = reqAny.(*http.Request)
 		}
```
Comment (bart0sh) [inline pkg/kubelet/prober/prober.go:162]:
```suggestion
                req = reqAny.(*http.Request)
		if !found || (p.HTTPGet.Host == "" && req.URL.Hostname() != status.PodIP) {
			req, err = httpprobe.NewRequestForHTTPGetAction(p.HTTPGet, &container, status.PodIP, "probe")
			if err != nil {
				return probe.Unknown, "", err
			}
			httpProbeReqCache.Store(key, req)
		}
```

Comment (Taoup) [reply pkg/kubelet/prober/prober.go:162]: Sorry, as a newbie in go, I'm not sure what will happen in case of a cache miss, reqAny will be nil? Is it ok to do a type assertion in that case?

Comment (bart0sh) [reply pkg/kubelet/prober/prober.go:162]: Good question. Go will panic trying to convert nil interface to *http.Request, I guess. My bad.

Comment (bart0sh) [reply pkg/kubelet/prober/prober.go:162]: I've updated the suggestion.

Comment (Taoup) [reply pkg/kubelet/prober/prober.go:162]: this is really neat. I'll make those changes

File: pkg/kubelet/prober/prober.go (hunk 2 above)
Comment (Taoup) [reply pkg/kubelet/prober/prober.go:46]: I tried to move the cache to httpprober. However, that prevent me from reusing the probeKey struct in the prober package. While move the cache to prober struct is quite simple and straightforward.

Comment (aojea) [reply pkg/kubelet/prober/prober.go]: we have bad experience caching globally because they can create memory leaks if some objects are never deleted , and those bugs were very subtle to catch. I think we should avoid creating a cache that is global

**Comment by aojea** [comment]
> update benchmark result:
> 
> ```
> benchstat old.txt new.txt 
> goos: linux
> goarch: amd64
> pkg: k8s.io/kubernetes/pkg/kubelet/prober
> cpu: Intel(R) Xeon(R) Gold 5220R CPU @ 2.20GHz
>         │   old.txt   │               new.txt               │
>         │   sec/op    │   sec/op     vs base                │
> Probe-8   557.6m ± 5%   164.7m ± 6%  -70.47% (p=0.000 n=10)
> 
>         │   old.txt    │               new.txt                │
>         │     B/op     │     B/op      vs base                │
> Probe-8   359.9Mi ± 0%   142.2Mi ± 0%  -60.49% (p=0.000 n=10)
> 
>         │   old.txt    │               new.txt               │
>         │  allocs/op   │  allocs/op   vs base                │
> Probe-8   3466.0k ± 0%   693.2k ± 0%  -80.00% (p=0.000 n=10)
> ```

can you add the benchmark to the code and reflect the results in a comment?

**Comment by Taoup** [comment]
> > update benchmark result:
> > ```
> > benchstat old.txt new.txt 
> > goos: linux
> > goarch: amd64
> > pkg: k8s.io/kubernetes/pkg/kubelet/prober
> > cpu: Intel(R) Xeon(R) Gold 5220R CPU @ 2.20GHz
> >         │   old.txt   │               new.txt               │
> >         │   sec/op    │   sec/op     vs base                │
> > Probe-8   557.6m ± 5%   164.7m ± 6%  -70.47% (p=0.000 n=10)
> > 
> >         │   old.txt    │               new.txt                │
> >         │     B/op     │     B/op      vs base                │
> > Probe-8   359.9Mi ± 0%   142.2Mi ± 0%  -60.49% (p=0.000 n=10)
> > 
> >         │   old.txt    │               new.txt               │
> >         │  allocs/op   │  allocs/op   vs base                │
> > Probe-8   3466.0k ± 0%   693.2k ± 0%  -80.00% (p=0.000 n=10)
> > ```
> 
> can you add the benchmark to the code and reflect the results in a comment?


the code doing this benchmark: 4c0578c9c8e19ee1d393052b3d573e754f3a1ac0

```bash
goos: linux
goarch: amd64
pkg: k8s.io/kubernetes/pkg/kubelet/prober
cpu: Intel(R) Xeon(R) Gold 5220R CPU @ 2.20GHz
            │    old.txt    │               new.txt               │
            │    sec/op     │   sec/op     vs base                │
HTTPProbe-8   142.71m ± 10%   33.33m ± 5%  -76.65% (p=0.000 n=10)

            │    old.txt    │               new.txt                │
            │     B/op      │     B/op      vs base                │
HTTPProbe-8   251.43Mi ± 0%   35.01Mi ± 0%  -86.08% (p=0.000 n=10)

            │   old.txt    │               new.txt               │
            │  allocs/op   │  allocs/op   vs base                │
HTTPProbe-8   3367.7k ± 0%   607.0k ± 0%  -81.98% (p=0.000 n=10)
```

**Comment by aojea** [comment]
> the code doing this benchmark: https://github.com/kubernetes/kubernetes/commit/4c0578c9c8e19ee1d393052b3d573e754f3a1ac0

yeah, but add those results  to the commit message and/or as a comment on top of the benchmark function, so we have a place where we can look in the code

**Comment by Taoup** [comment]
> > the code doing this benchmark: [4c0578c](https://github.com/kubernetes/kubernetes/commit/4c0578c9c8e19ee1d393052b3d573e754f3a1ac0)
> 
> yeah, but add those results to the commit message and/or as a comment on top of the benchmark function, so we have a place where we can look in the code

sry I misunderstood your intent, updated the commit.

File: pkg/kubelet/prober/prober.go (hunk 2 above)
Comment (bart0sh) [reply pkg/kubelet/prober/prober.go]: > I tried to move the cache to httpprober. However, that prevent me from reusing the probeKey struct in the prober package. While move the cache to prober struct is quite simple and straightforward.

Well, as caching makes sense only for HTTP probes, the logical place for it is the `httpprober` from my point of view. Reusing probeKey shouldn't stop you from at least trying. You can either move it somewhere (to avoid cyclical import) and make it public, or even duplicating in in the `httpprober`, which is also ok from my point of view.

I'd suggest to look at the possibility to cache http requests in the`httpprobe.NewRequestForHTTPGetAction` and may be renaming this method. This would allow you to cache [lifecycle probes](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/lifecycle/handlers.go#L159) as well.

File: pkg/kubelet/prober/prober.go (hunk 4 above)
Comment (bart0sh) [reply pkg/kubelet/prober/prober.go]: Frankly, It doesn't look very readable to me. I'd even try to avoid using sync.Map because of this and use plain map + locks (if locks are really needed here).

File: pkg/kubelet/prober/prober.go [hunk 5]
```diff
@@ -142,9 +146,16 @@ func (pb *prober) runProbe(ctx context.Context, probeType probeType, p *v1.Probe
...
 
 	case p.HTTPGet != nil:
-		req, err := httpprobe.NewRequestForHTTPGetAction(p.HTTPGet, &container, status.PodIP, "probe")
-		if err != nil {
-			return probe.Unknown, "", err
+		var err error
+		key := probeKey{podUID: pod.UID, containerName: container.Name, probeType: probeType}
+		reqAny, found := pb.httpProbeReqCache.Load(key)
+		req, _ := reqAny.(*http.Request)
+		if !found || (p.HTTPGet.Host == "" && req.URL.Hostname() != status.PodIP) {
```
Comment (bart0sh) [inline pkg/kubelet/prober/prober.go:153]: This is not very obvious (at least to me). Can you add a comment explaining what you're doing here and why?

Comment (Taoup) [reply pkg/kubelet/prober/prober.go:153]: I added this logic because I knew from many places that a pod is mutable.

As https://github.com/kubernetes/kubernetes/issues/108281#issuecomment-1058503524 states,  
> I don't think we are really saying that the IP can change underneath a running pod.
I'm considering removing this logic totally. If IP for a running pod won't change, can I assume the http probe request for a running pod won't stale?

File: pkg/kubelet/prober/prober.go (hunk 2 above)
Comment (Taoup) [reply pkg/kubelet/prober/prober.go]: > Well, as caching makes sense only for HTTP probes, the logical place for it is the httpprober from my point of view. Reusing probeKey shouldn't stop you from at least trying. You can either move it somewhere (to avoid cyclical import) and make it public, or even duplicating in in the httpprober, which is also ok from my point of view.

I agree that the logical place for the cache should be in httpprober. If you prefer, I could implement the cache in httpprober to make a comparison.

> I'd suggest to look at the possibility to cache http requests in thehttpprobe.NewRequestForHTTPGetAction and may be renaming this method. This would allow you to cache [lifecycle probes](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/lifecycle/handlers.go#L159) as well.

I read through the code and the docs, it seems that lifecycle hooks does not  get executed as frequently as normal probes, i.e. only upon container start and termination. I guess it does't worth a cache to do that.

File: pkg/probe/http/http.go [hunk 6]
```diff
@@ -90,7 +116,7 @@ type GetHTTPInterface interface {
 // If the HTTP response code is successful (i.e. 400 > code >= 200), it returns Success.
 // If the HTTP response code is unsuccessful or HTTP communication fails, it returns Failure.
 // This is exported because some other packages may want to do direct HTTP probes.
-func DoHTTPProbe(req *http.Request, client GetHTTPInterface) (probe.Result, string, error) {
+var DoHTTPProbe = func(req *http.Request, client GetHTTPInterface) (probe.Result, string, error) {
```
Comment (bart0sh) [inline pkg/probe/http/http.go:119]: Can you explain why is this change needed?

File: pkg/kubelet/prober/prober.go [hunk 7]
```diff
@@ -142,7 +145,7 @@ func (pb *prober) runProbe(ctx context.Context, probeType probeType, p *v1.Probe
 		return pb.exec.Probe(pb.newExecInContainer(ctx, container, containerID, command, timeout))
 
 	case p.HTTPGet != nil:
-		req, err := httpprobe.NewRequestForHTTPGetAction(p.HTTPGet, &container, status.PodIP, "probe")
+		req, err := pb.http.(*httpprobe.HttpProber).GetRequestForHTTPGetAction(p.HTTPGet, &container, status.PodIP, "probe")
```
Comment (bart0sh) [inline pkg/kubelet/prober/prober.go:148]: Can this casting be avoided? It makes the code less clear.

File: pkg/kubelet/prober/prober.go [hunk 8]
```diff
@@ -49,6 +50,8 @@ type prober struct {
 	runner kubecontainer.CommandRunner
 
 	recorder record.EventRecorder
+
+	httpProbeReqCache sync.Map
```
Comment (bart0sh) [inline pkg/kubelet/prober/prober.go:54]: This should be moved to the http probe code, I believe.

File: pkg/probe/http/http.go [hunk 9]
```diff
@@ -58,21 +60,22 @@ func NewWithTLSConfig(config *tls.Config, followNonLocalRedirects bool) Prober {
...
 // Prober is an interface that defines the Probe function for doing HTTP readiness/liveness checks.
 type Prober interface {
 	Probe(req *http.Request, timeout time.Duration) (probe.Result, string, error)
 }
 
-type httpProber struct {
+type HttpProber struct {
 	transport               *http.Transport
 	followNonLocalRedirects bool
+	httpRequestCache        *sync.Map
```
Comment (bart0sh) [inline pkg/probe/http/http.go:74]: Why this is not enough and the same field is added to the `prober` struct?

Comment (Taoup) [reply pkg/probe/http/http.go:74]: > Why this is not enough and the same field is added to the `prober` struct?

Oh, after a bunch of rebase things, I forgot the original one is still in `probe` struct.  Will delete it.

File: pkg/probe/http/http.go (hunk 6 above)
Comment (Taoup) [reply pkg/probe/http/http.go:119]: I changed this for the convenience of running the benchmark.  In the benchmark, this func variable is changed to a function always return true to mock the behavior of doing an actual http probe.  If this pr is to be merged, I think this change and the associated benchmark code is not necessary to be included.

File: pkg/kubelet/prober/prober.go (hunk 8 above)
Comment (Taoup) [reply pkg/kubelet/prober/prober.go:54]: sure
